designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/DBIOActionRunner.scala

```diff
@@ -18,6 +19,20 @@ class DBIOActionRunner(dbRef: DbRef) {
 
   def run[T](action: DB[T]): Future[T] =
     dbRef.db.run(action)
+
+  def safeRunInTransaction[T, E](action: DB[T])(error: Throwable => E)(
+      implicit ec: ExecutionContext
+  ): Future[Either[E, T]] =
+    safeRun(action.transactionally)(error)
+
+  def safeRun[E, T](action: DB[T])(error: Throwable => E)(
+      implicit ec: ExecutionContext
+  ): Future[Either[E, T]] =
+    run(action.map(t => Right(t)))
+      .recover { case ex: SQLException =>
```

Comment:
TBH we're cheating a little bit here. We catch `SQLException` but in the `error` conversion we have a `Throwable`. We do hide the SqlException, but in the end, we give the user a way to do something with it. 

Ideally, there should be no SQLException in the domain/core layer. It should be encapsulated in the repository. Maybe we can try to do it?

Currently we use in to run this code:
```scala
    val dbResult = for {
      dbFingerprint <- fingerprintRepository.read()
      result <- dbFingerprint match {
        case Some(dbValue) => DBIO.successful(new Fingerprint(dbValue))
        case None =>
          val generated = readFingerprintFromFile(fingerprintFileName).getOrElse(randomFingerprint)
          logger.info(s"Generated fingerprint $generated")
          fingerprintRepository
            .write(generated)
            .map(_ => new Fingerprint(generated))
      }
    } yield result
```

Are we able to add a new method to the `FingerprintRepository` interface? 
I'd see its declaration like this:
```scala
def readOrSave(freshFingerprint: Fingerprint): F[Either[Error, Fingerprint]]

sealed trait Error
object Error {
  case object SaveError extends Error
}
```

The code I show at the beginning (after some tweaks) would be its implementation. The SQLException will be caught in the repository implementation. At the core level, we should not bother of the underlying exceptions and just bubble them up. The error, we are interested in, is passed in the Left. 

WDYT?

Reply:
Done